File: app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/action/AdjustPlaybackSpeedAction.java
package org.jellyfin.androidtv.ui.playback.overlay.action;

import android.content.Context;
import android.view.View;
import android.widget.PopupWindow;

import org.jellyfin.androidtv.R;
import org.jellyfin.androidtv.ui.AudioDelayPopup;
import org.jellyfin.androidtv.ui.PlaybackSpeedPopup;
import org.jellyfin.androidtv.ui.ValueChangedListener;
import org.jellyfin.androidtv.ui.playback.PlaybackController;
import org.jellyfin.androidtv.ui.playback.overlay.CustomPlaybackTransportControlGlue;
import org.jellyfin.androidtv.ui.playback.overlay.LeanbackOverlayFragment;

public class AdjustPlaybackSpeedAction extends CustomAction {

    public AdjustPlaybackSpeedAction(Context context, CustomPlaybackTransportControlGlue customPlaybackTransportControlGlue) {
        super(context, customPlaybackTransportControlGlue);
        initializeWithIcon(R.drawable.exo_ic_speed);
    }

    @Override
    public void handleClickAction(PlaybackController playbackController, LeanbackOverlayFragment leanbackOverlayFragment, Context context, View view) {
        PlaybackSpeedPopup audioDelayPopup = new PlaybackSpeedPopup(context, view, new ValueChangedListener<Float>() {
            @Override
            public void onValueChanged(Float value) {
                playbackController.setPlaybackSpeed(value);
            }
        });

        PopupWindow popupWindow = audioDelayPopup.getPopupWindow();
        if (popupWindow != null) {
            popupWindow.setOnDismissListener(() -> {
                leanbackOverlayFragment.setFading(true);
            });
        }

        audioDelayPopup.show(playbackController.getPlaybackSpeed());
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/overlay/action/AdjustPlaybackSpeedAction.kt
package org.jellyfin.androidtv.ui.playback.overlay.action

import android.content.Context
import android.view.View
import org.jellyfin.androidtv.ui.playback.overlay.CustomPlaybackTransportControlGlue
import org.jellyfin.androidtv.ui.playback.PlaybackController
import org.jellyfin.androidtv.ui.playback.overlay.LeanbackOverlayFragment
import org.jellyfin.androidtv.ui.PlaybackSpeedPopup
import org.jellyfin.androidtv.R

class AdjustPlaybackSpeedAction(
    context: Context,
    customPlaybackTransportControlGlue: CustomPlaybackTransportControlGlue
) : CustomAction(context, customPlaybackTransportControlGlue) {
    override fun handleClickAction(
        playbackController: PlaybackController,
        leanbackOverlayFragment: LeanbackOverlayFragment,
        context: Context,
        view: View
    ) {
        val playbackSpeedPopup = PlaybackSpeedPopup(context, view) { value: Float ->
			playbackController.playbackSpeed = value
		}
		playbackSpeedPopup.popupWindow?.setOnDismissListener { leanbackOverlayFragment.setFading(true) }
        playbackSpeedPopup.show(playbackController.playbackSpeed)
    }

    init {
        initializeWithIcon(R.drawable.exo_ic_speed)
    }
}
